src/shm_sub.c

```diff
@@ -1606,18 +1606,30 @@ sr_shmsub_change_notify_change(struct sr_mod_info_s *mod_info, const char *orig_
             nsub->pending_event = 1;
             pending_events = 1;
 
-            /* prepare the diff to write into subscription SHM */
-            if ((err_info = sr_shmsub_change_notify_get_diff(mod_info->diff, nsub->mod->ly_mod, opts, &full_diff_lyb,
-                    &full_diff_lyb_len, &diff_lyb, &diff_lyb_len, &free_diff))) {
-                goto cleanup;
-            }
-
             /* open sub SHM and map it */
             if ((err_info = sr_shmsub_open_map(nsub->mod->ly_mod->name, sr_ds2str(mod_info->ds), -1, &nsub->shm_sub))) {
                 goto cleanup;
             }
             nsub->sub_shm = (sr_sub_shm_t *)nsub->shm_sub.addr;
 
+            if (nsub->mod->reuse_diff && !(opts & SR_SUBSCR_CHANGE_ALL_MODULES)) {
+                /* if we have prepared a module diff before and we don't need all modules this time */
+                if (free_diff) {
+                    free(diff_lyb);
+                }
+                diff_lyb = NULL;
+                diff_lyb_len = 0;
+            } else {
+                /* prepare the diff to write into subscription SHM */
+                if ((err_info = sr_shmsub_change_notify_get_diff(mod_info->diff, nsub->mod->ly_mod, opts, &full_diff_lyb,
+                        &full_diff_lyb_len, &diff_lyb, &diff_lyb_len, &free_diff))) {
+                    goto cleanup;
+                }
+
+                /* set mod->reuse_diff only if we prepared a per-module diff */
+                nsub->mod->reuse_diff = !(opts & SR_SUBSCR_CHANGE_ALL_MODULES);
+            }
+
```

Comment:
Not sure why you moved the code generating the `diff` after sub SHM open but I suppose it makes no difference. But please move all this new logic into `sr_shmsub_change_notify_get_diff`, no reason not to and it will not make these notify functions longer, they already are long enough.

Reply:
I moved the code generating diff below sub shm open, because earlier I was also storing the diff_lyb_len in the shm, to pass onto `sr_shmsub_notify_write_event` to prevent any accidental remaps.

However, I realized that this was not necessary, because xpath is always NULL for change_sub events. So as long as the diff_lyb_len remains 0, we should not have a problem.

But it may still make sense to write the diff_lyb_len into the SHM for a paranoid/sanity check?